chain-abci/src/enclave_bridge/edp/tdbe.rs

```diff
@@ -0,0 +1,238 @@
+    /// External TDBE server address, used by remote nodes to send RPC requests. E.g.
+    /// `<public_ip>:<public_port>`
+    pub external_listen_address: String,
+}
+
+#[derive(Debug)]
+pub struct TdbeApp {
+    /// UDS to connect to `chain-abci`
+    chain_abci_stream: UnixStream,
+    /// UDS to persist data to `chain-storage`
+    persistence_stream: UnixStream,
+    /// `ra-sp-server` address for remote attestation. E.g. `0.0.0.0:8989`
+    /// TODO:  Replace it with a local UDS (using `chain-abci` as launcher).
+    sp_address: String,
+    /// Optional address of remote TDBE server for fetching initial data
+    remote_tdbe_address: Option<String>,
+    /// Optional DNS name of remote TDBE server for fetching initial data (`chain-abci` fetches it
+    /// from RPC and passes it to TDBE App)
+    remote_tdbe_dns_name: Option<String>,
+    /// Local TDBE server address to listen on. E.g. `127.0.0.1:3445`
+    local_listen_address: String,
+}
+
+impl TdbeApp {
+    /// Creates a new instance of TDBE app
+    pub fn new(
+        tdbe_config: &TdbeConfig,
+        ra_config: &SpRaConfig,
+        storage: Arc<dyn KeyValueDB>,
+    ) -> std::io::Result<Self> {
+        let (chain_abci_stream, chain_abci_receiver) = UnixStream::pair()?;
+        let (persistence_stream, persistence_receiver) = UnixStream::pair()?;
+        let (remote_tdbe_address, remote_tdbe_dns_name) =
+            fetch_remote_tdbe_connection_details(tdbe_config)?;
+
+        spawn_chain_abci_thread(chain_abci_receiver, storage.clone());
+        spawn_persistence_thread(persistence_receiver, storage);
+
+        Ok(Self {
+            chain_abci_stream,
+            persistence_stream,
+            sp_address: ra_config.address.clone(),
+            remote_tdbe_address,
+            remote_tdbe_dns_name,
+            local_listen_address: tdbe_config.local_listen_address.clone(),
+        })
+    }
+
+    pub fn spawn(self) {
+        thread::spawn(move || {
+            let mut device = Device::new()
+                .expect("SGX device was not found")
+                .einittoken_provider(AesmClient::new())
+                .build();
+            let mut enclave_builder = EnclaveBuilder::new("tdb-enclave-app.sgxs".as_ref());
+
+            enclave_builder
+                .coresident_signature()
+                .expect("Enclave signature file not found");
+            enclave_builder.usercall_extension(self);
+
+            let enclave = enclave_builder
+                .build(&mut device)
+                .expect("Failed to build enclave");
+            enclave.run().expect("Failed to start enclave")
+        });
+    }
+}
+
+/// Fetches connection details of remote TDBE server using TM RPC
+fn fetch_remote_tdbe_connection_details(
+    _tdbe_config: &TdbeConfig,
+) -> std::io::Result<(Option<String>, Option<String>)> {
+    // TODO: Fetch connection details using TM RPC (use `client-common::TendermintRpcClient`?)
+    Ok((None, None))
+}
```

Comment:
this should happen inside the enclave -- so just need to expose the external node's RPC address to the enclave

Reply:
Ok. I'll modify.

Reply:
Changed.